semantic-tags/src/Tags/Taggable/Precise.hs

```diff
@@ -0,0 +1,147 @@
+{-# LANGUAGE AllowAmbiguousTypes, DataKinds, DeriveGeneric, DisambiguateRecordFields, FlexibleContexts, FlexibleInstances, MultiParamTypeClasses, NamedFieldPuns, OverloadedStrings, ScopedTypeVariables, TypeApplications, TypeFamilies, TypeOperators, UndecidableInstances #-}
+{-# LANGUAGE StandaloneDeriving #-}
+module Tags.Taggable.Precise
+( runTagging
+) where
+
+import           Control.Effect.Reader
+import           Data.Aeson as A
+import           Data.Monoid (Ap(..), Endo(..))
+import           Data.List.NonEmpty (NonEmpty(..))
+import           Data.Location
+import           Data.Source
+import           Data.Text as T
+import           GHC.Generics
+import qualified TreeSitter.Python.AST as Python
+
+data Tag = Tag
+  { name :: Text
+  , kind :: Kind
+  , span :: Span
+  , context :: [Text]
+  , line :: Maybe Text
+  , docs :: Maybe Text
+  }
+  deriving (Eq, Generic, Show)
+
+instance ToJSON Tag
+
+
+data Kind
+  = Function
+  | Method
+  | Class
+  | Module
+  | Call
+  deriving (Bounded, Enum, Eq, Generic, Show)
+
+instance ToJSON Kind where
+  toJSON = toJSON . show
+  toEncoding = toEncoding . show
+
+
+type ContextToken = (Text, Maybe Range)
+
+runTagging :: Source -> Python.Module Location -> [Tag]
+runTagging source
+  = ($ [])
+  . appEndo
+  . run
+  . runReader @[ContextToken] []
+  . runReader source
+  . tag where
+
+class ToTag t where
+  tag
+    :: ( Carrier sig m
+       , Member (Reader Source) sig
+       , Member (Reader [ContextToken]) sig
+       )
+    => t
+    -> m (Endo [Tag])
```

Comment:
Would it be more morally correct to pull in `list` and make this `m (DList Tag)`? Then you get an `IsList` instance and all that jazz.

Reply:
We definitely don’t want an `IsList` instance—it would be an error to generate multiple tags at a time—and in fact to as great a degree as possible nothing should even know about this type; e.g. I’ve since switched to using a `Writer` effect to accumulate these and `yield` tags explicitly.